File: capy/src/e2e-tests/java/dev/capylang/test/RecFunctionsTest.java
package dev.capylang.test;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class RecFunctionsTest {
    @Test
    void tailRecursiveTopLevelFunctionWorks() {
        assertThat(RecFunctions.sum(5, 0)).isEqualTo(15);
    }

    @Test
    void tailRecursiveLocalFunctionWorks() {
        assertThat(RecFunctions.sumInner(5)).isEqualTo(15);
    }

    @Test
    void tailRecursiveFunctionSupportsLetAndMatchTailPositions() {
        assertThat(RecFunctions.sumWithLet(5, 0)).isEqualTo(15);
        assertThat(RecFunctions.sumWithMatch(5, 0)).isEqualTo(15);
    }

    @Test
    void tailRecursiveFunctionIsLoweredWithoutJavaStackGrowth() {
        assertThat(RecFunctions.countDownLarge(50_000, 0)).isEqualTo(50_000);
    }

    @Test
    void tailRecursiveMatchPreservesOptionAndResultBindings() {
        assertThat(RecFunctions.sumPresentValues(List.of(1, 2, 3), 0)).isEqualTo(6);
        assertThat(RecFunctions.sumLengthsUntilError(List.of("a", "bb", "bad", "cccc"), 0))
                .isEqualTo("error:bad length:3");
        assertThat(RecFunctions.sumLengthsUntilError(List.of("a", "bb", "ccc"), 0)).isEqualTo("ok:6");
    }
}

File: capy/src/e2e-cfun/js/dev/capylang/test/RecFunctions.test.js
const test = require('node:test');
const { assert, generatedModule } = require('./_support');

test('RecFunctions', () => {
    const rec = generatedModule('dev/capylang/test/RecFunctions.js');
    assert.equal(rec.sum(5, 0), 15);
    assert.equal(rec.sumInner(5), 15);
    assert.equal(rec.sumWithLet(5, 0), 15);
    assert.equal(rec.factorialLarge(5, 1), 120);
    assert.equal(rec.unmarkedTailSum(5, 0), 15);
    assert.equal(rec.unmarkedLocalTailFactorialLarge(5), 120);
    assert.equal(rec.unmarkedNonTailSum(5), 15);
    assert.equal(rec.sumPresentValues([1, 2, 3], 0), 6);
    assert.equal(rec.sumLengthsUntilError(['a', 'bb', 'bad', 'cccc'], 0), 'error:bad length:3');
    assert.equal(rec.sumLengthsUntilError(['a', 'bb', 'ccc'], 0), 'ok:6');
});
